feat: Telescope picker

al-ce opened the pull request:
Copying the commit message here for convenience:

---

`:Hbac telescope` or `require("hbac").telescope()` spawns a Telescope picker of listed buffers with pin state indicators in the first column. Icons for the pin state and mappings for the picker's commands have defaults that can be overridden in the setup config. The following picker commands are available:

- `toggle_selections` toggles the pin state of selected buffers (either single or multi select)
- `toggle_all` calls the `toggle_all` subcommand (also available as the new `Hbac toggle_all`), which sets the pin state of all buffers to the current `state.pin_all` value and then flips that value
- `close_unpinned` calls the `close_unpinned` subcommand, which closes all unpinned buffers except those visible in windows. Previously, `Hbac close_unpinned` would close unpinned buffers in windows. This has been fixed per @axkirillov's suggestion
- `delete_buffer` deletes selected buffers using `opts.close_command`

After each command, the picker is refreshed and the pin icons are updated. Pressing the default action key (typically enter) opens the selected buffer in the current window.

This commit updates the module structure in order to avoid circular imports between the `command` and `telescope` modules. The `command` module is now in its own directory. Both the main `command` module and the `telescope` module call command functions from `command.subcommands`. In order to avoid displaying notifications of Hbac events while Telescope commands are executed, the notifications have been removed from the main subcommand functions and moved into the main `command` module. The notifications have also been changed from `print` to `vim.notify` functions.

The refactoring and renaming of files changes code made by @axkirillov and @ditsuke.

---

al-ce committed:
refactor: Move buflist generator to new utils module

- Standardizes the checks used to generate the filtered buflist used across a few modules

al-ce committed:
feat: Add Telescope picker for managing buf pin state

Adds a Telescope picker that displays all listed buffers and their pin status, indicated by a customizable icon. The following actions can be performed on any buffer(s) selected in the picker:

- `<M-y>`: Toggle the pin status of all selected buffers. Multiple buffers can be selected (default `<Tab>`). After the toggle action, the icon will be updated to reflect the new pin status. The picker does not close.
- `<M-a>`: Same as above, but pins/unpins all listed buffers
- `<M-x>`: Delete all selected buffers. The picker does not close. Uses the function defined in `opts.close_command`
- `<M-c>`: Close all unpinned buffers. The picker does not close.
- `<CR>`: Opens the selected buffer in the current window and closes the picker.

The picker depends on `telescope.nvim` and its dependency `plenary.nvim`. A `pcall` check is used to prevent the module from loading if these are not present.
The display style of the filepath in the picker is copied from [smart-open.nvim](https://github.com/danielfalk/smart-open.nvim) by @danielfalk.

The `opts` table adds configurable options for the picker in the `telescope` table:
- `mappings`: set keybindings for the actions described above
- `pin_icons`: set the pin/unpin icons and their respective highlight groups

Co-Authored-By: Aleksandr Kirillov <32141102+axkirillov@users.noreply.github.com>
Co-Authored-By: Tushar <hello@ditsuke.com>

al-ce commented on lua/hbac/setup.lua:
```diff
@@ -8,6 +8,17 @@ local M = {
 		close_command = function(bufnr)
 			vim.api.nvim_buf_delete(bufnr, {})
 		end,
+		telescope = {
+			mappings = {
+				-- first key is for "normal", second for "insert" mode
+				close_unpinned = { "<C-,>", "<C-,>" },
+				delete_buffer = { "<C-.>", "<C-.>" },
+				toggle_all = { "<C-;>", "<C-;>" },
+				toggle_selections = { "<C-g>", "<C-g>" },
+			},
+			-- pinned icon, unpinned icon, pinned hl, unpinned hl
+			pin_icons = { "󰐃", "", "DiagnosticError", "Normal" },
+		},
```
Was looking for non-conflicting default mappings but these are admittedly awkward. Could also just leave them blank and require a user to set them. Pin icons are nerd font icons.

I made all these positional to avoid extra keys and more nested tables. Please advise on these choices.

axkirillov replied:
You can also use the meta key (M) for mappings to avoid conflicts.

al-ce commented on lua/hbac/state.lua:
```diff
@@ -1,16 +1,12 @@
-
-	M.pinned_buffers[bufnr] = true
-	return "pinned"
+	M.pinned_buffers[bufnr] = not M.pinned_buffers[bufnr]
+	return M.pinned_buffers[bufnr]
```
This is part of the overall shift to letting the main `command` module handle notifications.

al-ce commented on lua/hbac/telescope/init.lua:
```diff
@@ -0,0 +1,93 @@
+	return true
+end
+
+M.pin_picker = function(opts)
+	opts = opts or {}
+	pickers
+		.new(opts, {
+			prompt_title = "Hbac Pin States",
+			finder = make_finder(),
+			sorter = telescope_conf.generic_sorter(opts),
+			attach_mappings = attach_mappings,
+			previewer = telescope_conf.file_previewer(opts),
+		})
+		:find()
+end
```
I wasn't sure how to interpret the request to make the picker more like the buffer picker beyond letting the default action (enter) open the selected buffer in a window. This is what my buffer picker looks like:

<img width="800" alt="SCR-20230515-tfdt" src="https://github.com/axkirillov/hbac.nvim/assets/23170004/3559d6df-a835-4cba-85fc-4ba529e163d6">

This is my first time making a Telescope picker so any advice and criticism is appreciated. What other changes would you like to see?

al-ce commented on lua/hbac/utils.lua:
```diff
@@ -0,0 +1,14 @@
+M = {}
+
+M.get_buffers = function()
+	return vim.tbl_filter(function(bufnr)
+		local name = vim.api.nvim_buf_get_name(bufnr)
+		local listed = vim.api.nvim_buf_get_option(bufnr, "buflisted")
+		if name ~= "" and listed then
+			return true
+		end
+		return false
+	end, vim.api.nvim_list_bufs())
+end
```
This is the filter function from @ditsuke's PR that I wanted to use to standardize the buflist. Aside from the additional check to see if a buffer is in a window, is there any place we shouldn't be using this?

al-ce commented on lua/hbac/command/subcommands.lua:
```diff
@@ -0,0 +1,45 @@
+	local buflist = utils.get_buffers()
+	for _, bufnr in ipairs(buflist) do
+		local windows_with_buf = vim.fn.win_findbuf(bufnr)
+		if bufnr ~= curbufnr and not state.is_pinned(bufnr) and #windows_with_buf == 0 then
+			hbac_config.close_command(bufnr)
+		end
```
Not a fan of the conjunct triplet here but maybe that's nitpicky

axkirillov replied:
It's either this or extracting it to a function with a series of ifs 🤷🏻

al-ce commented:
If this PR is too big, we can close it and start smaller! Feel free to push any changes big or small as you see fit. Thank you!

https://github.com/al-ce/hbac.nvim/assets/23170004/d9e6f03d-451b-455e-be84-caf112ff5595

axkirillov commented on lua/hbac/command/subcommands.lua:
```diff
@@ -0,0 +1,45 @@
+	return bufnr, pinned_state
+end
+
+M.pin_all = function()
```
I guess this is toggle_all because it works both ways, but I'm ok with pin_all as well

al-ce replied:
Yes my mistake, I forgot to change `toggle_all` to `pin_all` in the subcommands module. Both names are kind of misleading though (pin_all doesn't invite the user to try to unpin all and toggle_all suggests toggling the state of all buffers to their opposite rather than all to pinned/unpinned). Naming things is hard.

axkirillov replied:
I thought of the same thing, so I am fine with either.

axkirillov commented on lua/hbac/command/subcommands.lua:
```diff
@@ -0,0 +1,45 @@
+end
+
+M.pin_all = function()
+	local buflist = utils.get_buffers()
+	for _, bufnr in ipairs(buflist) do
+		state.pinned_buffers[bufnr] = state.pin_all
+	end
+	state.pin_all = not state.pin_all
+	return state.pin_all
```
This probably could work as a module scoped side effect, since it's unlikely other modules would reach for this piece of state
```suggestion
	local buflist = utils.get_buffers()
	for _, bufnr in ipairs(buflist) do
		state.pinned_buffers[bufnr] = M.pin_all
	end
	M.pin_all = not M.pin_all
	return M.pin_all
```

al-ce replied:
Good point. So, move `state.pin_all` to this module and do some renaming. I'm thinking we should go with `toggle_all` as the name for this function and the boolean value to something more obvious, maybe `M.toggle_all_value`.

axkirillov replied:
maybe something like `set_all_to`? The wrapping function performs a toggle indeed, but on the inside what you really do it set all values to some value, so maybe it can reflect that?

al-ce replied:
That sounds good, am I interpreting this correctly?

```lua
local M = {
	set_all_to_value = true,
}

M.set_all_to = function()
	local buflist = utils.get_buffers()
	for _, bufnr in ipairs(buflist) do
		state.pinned_buffers[bufnr] = M.set_all_to_value
	end
	M.set_all_to_value = not M.set_all_to_value
	return M.set_all_to_value
end
```

axkirillov replied:
not quite, I was only talking about the variable name, the function name is still a toggle so
```lua
local M = {
	set_all_to_value = true,
}

M.toggle_all = function()
	local buflist = utils.get_buffers()
	for _, bufnr in ipairs(buflist) do
		state.pinned_buffers[bufnr] = M.set_all_to_value
	end
	M.set_all_to_value = not M.set_all_to_value
	return M.set_all_to_value
end
```

axkirillov replied:
Now that I looked at both variants I think I like yours better. The one where the function is `toggle_all` and the variable `toggle_all_value` 😅 
Mine is kinda confusing cause it sounds like a method name

axkirillov commented on lua/hbac/command/subcommands.lua:
```diff
@@ -0,0 +1,45 @@
+end
+
+M.toggle_autoclose = function()
+	if state.autoclose_enabled then
+		state.autoclose_enabled = false
+		autocommands.autoclose.disable()
+	else
+		state.autoclose_enabled = true
+		autocommands.autoclose.setup()
+	end
+	return state.autoclose_enabled
```
I like early returns. What do you think? Is this more / less readable?
```suggestion
    state.autoclose_enabled = not state autoclose_enabled
	if state.autoclose_enabled then
		autocommands.autoclose.setup()
		return state.autoclose_enabled
	end
	autocommands.autoclose.disable()
	return state.autoclose_enabled
```

axkirillov commented on lua/hbac/telescope/make_display.lua:
```diff
@@ -0,0 +1,64 @@
+local state = require("hbac.state")
+
+local entry_display = require("telescope.pickers.entry_display")
+local has_devicons, devicons = pcall(require, "nvim-web-devicons")
```
since we require telescope, devicons and plenary, all of this should be mentioned in the README

axkirillov commented on lua/hbac/telescope/init.lua:
```diff
@@ -0,0 +1,93 @@
+	return true
+end
+
+M.pin_picker = function(opts)
```
I think we should print a nice message for ppl who don't have telescope installed, plenary also seems to be necessary

al-ce replied:
Should we use a `pcall` check, print the message, and return?

axkirillov replied:
yep, you already have a has_devicons in make_display, perhaps there could be a has_telescope here

al-ce commented on lua/hbac/command/init.lua:
```diff
@@ -0,0 +1,49 @@
+	vim.notify(bufnr .. " " .. pinned_state, "info", { title = "Hbac" })
+end
+
+M.subcommands.toggle_all = function()
+	local pin_all_state = subcommands.pin_all() and "pinned" or "unpinned"
+	vim.notify(pin_all_state .. " all buffers", "info", { title = "Hbac" })
+end
```
Per the other conversation regarding `toggle_all`, should probably rename this
```suggestion
M.subcommands.toggle_all = function()
	local toggle_all_value = subcommands.pin_all() and "pinned" or "unpinned"
	vim.notify(toggle_all_value .. " all buffers", "info", { title = "Hbac" })
end
```

We can sleep on these though 😆  (these moments remind me of the Greek philosopher Cratylus who allegedly gave up language at the end of his life and spoke only with his finger, since he found words too slippery and incomplete)

al-ce commented on lua/hbac/telescope/init.lua:
```diff
@@ -0,0 +1,93 @@
+local M = {}
```
```suggestion
if not pcall(require, "telescope") then
	local msg = "Please install telescope.nvim to use hbac's telescope picker"
	vim.notify(msg, vim.log.levels.ERROR, { title = "Hbac" })
	return
end

local M = {}
```

Since plenary is a dependency for telescope and it's only used as part of the telescope module, I didn't add it. Should it be there anyway?

al-ce commented on lua/hbac/telescope/init.lua:
```diff
@@ -0,0 +1,93 @@
+local M = {}
+
+local hbac_config = require("hbac.setup").opts
+local make_display = require("hbac.telescope.make_display")
+local state = require("hbac.state")
+local subcommands = require("hbac.command.subcommands")
+local utils = require("hbac.utils")
```
I'm noticing I need to move `hbac_config` to `attach_mappings` because otherwise, with the current design, the default opts are being called before the user opts are loaded. I was kind of reckless with just importing everything left and right and need to take a good look at it, but please keep a lookout for anything fishy or something that might cause future problems in this regard. I'm thinking of moving the other hbac module imports to `attach_mappings` too?

al-ce commented on lua/hbac/telescope/init.lua:
```diff
@@ -0,0 +1,100 @@
+if not pcall(require, "telescope") then
+	return false
+end
```
I didn't add a check for plenary since plenary is already a dependency for telescope, but I did mention it in the README.

ditsuke replied:
Let's be explicit?

al-ce replied:
@ditsuke something like

```lua
local dependencies = {
  "plenary",
  "telescope",
}

for _, dependency in ipairs(dependencies) do
  if not pcall(require, dependency) then
    return dependency
  end
end
```

So we can specify what's missing? Or did you mean something else?

ditsuke replied:
Yes, let's just be clear we're directly depending on plenary.

al-ce replied:
https://github.com/al-ce/hbac.nvim/commit/4c01fd575ea37d3c79fe318af37f5ba9031aabcc

How does this look @ditsuke? I moved the notification out of the function in `hbac.command`, it was going to get a bit messy.

ditsuke replied:
Looks good! Small nit, since the Telescope integration is optional it might be better to check those dependencies only if that component is invoked.

al-ce replied:
Thanks! I thought that was the case already, since as far as I can tell the telescope module is only required if a user executes `:Hbac telescope` or `require("hbac").telescope()`, but I may have missed some other spot?

axkirillov replied:
As far as I can see the telescope module is only required during the execution of telescope command, so it should be OK.

Furthermore the module itself should not be required unless the dependencies are set. So I think it is apropriate to check at the top of the module.

ditsuke replied:
My bad, looks good then 💯

axkirillov commented on README.md:
```diff
@@ -10,14 +10,27 @@ The main feature of this plugin, however, is the automatic closing of buffers. I
-use 'axkirillov/hbac.nvim'
+use {
+  'axkirillov/hbac.nvim',
+  requires = {
```
```suggestion
  requires = {
  -- these are optional, add them, if you want the telescope module
```

axkirillov commented on README.md:
```diff
@@ -10,14 +10,27 @@ The main feature of this plugin, however, is the automatic closing of buffers. I
 {
   'axkirillov/hbac.nvim',
+  -- For Telescope integration
+  dependencies = {
```
```suggestion
  dependencies = {
  -- these are optional, add them, if you want the telescope module
```

axkirillov commented on lua/hbac/setup.lua:
```diff
@@ -2,12 +2,21 @@ local state = require("hbac.state")
+		telescope = {
+			mappings = {
+				close_unpinned = { "<M-c>" },
+				delete_buffer = { "<M-d>" },
```
I like to use `<C-x>`, `<M-d>` is delete word forward in readline.

axkirillov commented:
I get
![image](https://github.com/axkirillov/hbac.nvim/assets/32141102/946069d5-a9fb-440c-9082-9fc342693c5a)

al-ce commented:
> I get

Sorry about that, must be something from this change.

https://github.com/al-ce/hbac.nvim/blob/9ad3d209fa5f22779b45219c237a544e6d48a091/lua/hbac/telescope/init.lua#L71-L82

But I'm having a hard time figuring out what I've got in my config that's suppressing that error. Can I please see your opts in your hbac config, if you have any? Does the error show up on vim startup or when you run a specific Hbac command?

axkirillov commented:
@al-ce I'm relying on defaults
```lua
	{
		'axkirillov/hbac.nvim',
		config = function()
			require("hbac").setup()
		end,
		branch = "test"
	}

```

axkirillov commented:
> > I get
> 
> Sorry about that, must be something from this change.
> 
> https://github.com/al-ce/hbac.nvim/blob/9ad3d209fa5f22779b45219c237a544e6d48a091/lua/hbac/telescope/init.lua#L71-L82
> 
> But I'm having a hard time figuring out what I've got in my config that's suppressing that error. Can I please see your opts in your hbac config, if you have any? Does the error show up on vim startup or when you run a specific Hbac command?

I think the first argument to map function is supposed to be string

al-ce commented:
> I think the first argument to map function is supposed to be string

I pushed that change, and using a table there was unnecessary anyway since the function is mapping one mode at a time regardless. Any change?

axkirillov commented:
> > I think the first argument to map function is supposed to be string
> 
> I pushed that change, and using a table there was unnecessary anyway since the function is mapping one mode at a time regardless. Any change?

Now it works, thanks.

One nitpick
![image](https://github.com/axkirillov/hbac.nvim/assets/32141102/a9f43989-cacc-4eba-8c65-46daee52b22a)
help should have a hint for those mappings

axkirillov commented:
https://github.com/axkirillov/hbac.nvim/assets/32141102/abb4ec26-f75f-49e2-9b2e-40fa35895e99

Using the toggle_selection resets the select position back to downmost entry. In case of multiple selection it makes sense, but if I am pinning one buffer at the top I expect the cursor / selection not to move
Perhaps it make sense to have a separate mapping for single buffer pinning? This must be some telescope mechanics that we can't influence, isn't it?

al-ce commented:
Thanks for the catch on the mapping issue!

> Using the toggle_selection resets the select position back to downmost entry. In case of multiple selection it makes sense, but if I am pinning one buffer at the top I expect the cursor / selection not to move

Yeah that's been bugging me too. Also I noticed that the 'unpinned' icon's bg highlight doesn't change when an unpinned buffer is selected, I didn't notice that with my colorscheme/highlights, I'd like to fix that.

al-ce commented:
> This must be some telescope mechanics that we can't influence, isn't it?

I think the current issue is that in order to update the pin status visual cue (the icons), the refresh function redraws the entire picker, but I'm hoping there's a way to store the row position and set it back. Navigating the Telescope codebase is kinda challenging but good exercise. This maybe looks promising? [set_selection](https://github.com/search?q=repo%3Anvim-telescope%2Ftelescope.nvim+set_selection&type=code) [Picker:set_selection(row)](https://github.com/nvim-telescope/telescope.nvim/blob/40c31fdde93bcd85aeb3447bb3e2a3208395a868/lua/telescope/pickers.lua#L935)

My next step is to dig through these [search results](https://github.com/search?q=picker%3Arefresh&type=code) for `picker:refresh` and see what others do in this scenario.

al-ce commented on lua/hbac/telescope/init.lua:
```diff
@@ -0,0 +1,119 @@
+	end
+
+
+	local hbac_toggle_selections = function(prompt_bufnr)
+		execute_telescope_action(prompt_bufnr, state.toggle_pin)
+	end
+	local hbac_toggle_all = function(prompt_bufnr)
+		execute_telescope_action(prompt_bufnr, subcommands.toggle_all)
+	end
+	local hbac_close_unpinned = function(prompt_bufnr)
+		execute_telescope_action(prompt_bufnr, subcommands.close_unpinned)
+	end
+	local hbac_delete_buffer = function(prompt_bufnr)
+		execute_telescope_action(prompt_bufnr, hbac_config.close_command)
+	end
+
+	actions.hbac_toggle_selections = hbac_toggle_selections
+	actions.hbac_toggle_all = hbac_toggle_all
+	actions.hbac_close_unpinned = hbac_close_unpinned
+	actions.hbac_delete_buffer = hbac_delete_buffer
```
To show up in the Telescope which_key popup, the functions need to be registered as named functions in `telescope.actions`, and previously I was just mapping them in a loop. This is kind of clunky but I'm having a hard time coming up with anything cleaner right now due to the named-function constraint. Need to stop staring at it for a while and maybe something will come. Appreciate any suggestions.

ditsuke replied:
How about

```lua
local actions = {
	hbac_delete_buffer = function() end,
	...
}

for name, action in pairs(actions) do
	actions[name] = action
done
```

al-ce replied:
@ditsuke I tried this initially but this registers them as anonymous functions so it goes back to displaying
<img src=https://user-images.githubusercontent.com/32141102/238931358-a9f43989-cacc-4eba-8c65-46daee52b22a.png>

I want to dig through the telescope code a little more so I can understand this, but so far I've only been able to get the which_key pop up hints to register with the same name they're assigned in the attach_mappings function. The `map()` function doesn't care what it gets, and that's why we can iterate through the mappings opt to set those. I'll keep looking though, I'm still not understanding the fundamentals of what's going on here.

axkirillov replied:
So this won't work right now? I'm getting the `<anonymous>` still. I guess we could create an issue and leave it for later, if you want.

al-ce replied:
@axkirillov sorry, something to do with the modularization of `attach_mappings` I did in 90689bec47bf24c782c9b1cd443ebf3bfd353825, will revert and go from there

al-ce replied:
Pushed [28fd951](https://github.com/axkirillov/hbac.nvim/pull/5/commits/28fd9513918af49fd555eceef4ac4bd6e5581ea9), how about now?

Telescope's [developers.md](https://github.com/nvim-telescope/telescope.nvim/blob/master/developers.md) mentions:

> We [replace the default action] by setting the attach_mappings key to a function. This function needs to return either true or false. If it returns false it means that only the actions defined in the function should be attached.

So I guess if I want to clean this up I need to figure out the constraints of "defined in the function". I'm also ok with leaving it as a task for someone more skilled :sweat_smile:

axkirillov replied:
It works now 👍🏻

ditsuke replied:
Works! :100:

al-ce commented on lua/hbac/setup.lua:
```diff
@@ -2,12 +2,21 @@ local state = require("hbac.state")
 		end,
+		telescope = {
+			mappings = {
+				close_unpinned = { "<M-c>" },
+				delete_buffer = { "<M-x>" },
+				toggle_all = { "<M-a>" },
+				toggle_selections = { "<M-y>" },
```
`<C-x>` is the Telescope default for 'Go to file selection as a split' so I hesitated to overwrite that. What about `<M-x>`?

ditsuke replied:
M-x seems reasonable

al-ce reviewed (commented):
I'm tentatively adopting the distinction between hbac telescope actions and hbac commands as of a4d6afd1c94c760475e0e95e5d5e0cdb16ef5f1c, hopefully it's clearer in the telescope module

axkirillov commented on lua/hbac/utils.lua:
```diff
@@ -0,0 +1,14 @@
+			return true
+		end
+		return false
+	end, vim.api.nvim_list_bufs())
```
Would you mind if I merge this first?
https://github.com/axkirillov/hbac.nvim/pull/8

Then this becomes
```lua
			local buffers = vim.tbl_filter(function(buf)
				return vim.api.nvim_buf_get_option(buf, "buflisted")
			end, vim.api.nvim_list_bufs())
``` 
We can then name it `get_listed_buffers`

al-ce replied:
Yep looks good! I was expecting this to change eventually to address #6.

al-ce replied:
@ditsuke do you mind pushing the fix for this here (or making a suggestion we can commit) and adding a call to this in `autocommands`? I feel bad having my name on the git blame for your work (this was already your code above..!)

ditsuke replied:
Done with [a30aa4a](https://github.com/axkirillov/hbac.nvim/pull/5/commits/a30aa4ac7c96a7f345e046567fe79154a18554f9)

al-ce commented:
Summary of last three commits
- e8a967a Move attach_mappings back to its own module, preserving the action hints in the which_key pop-up
- 5d4b8b1 Use direct declarations e.g. `local function funcname()` instead of local funcname = function()` except when functions are defined in a table. I found this more readable and it reflects the requirement that these functions need to be named in order to be mapped properly.
- 4c01fd5 Checks for missing plenary and telescope with pcall and notifies user of one or both
<img width="530" alt="SCR-20230519-qugz" src="https://github.com/al-ce/hbac.nvim/assets/23170004/0677bd30-ca5a-4818-a782-266322db8c8c">

ditsuke committed:
refactor(utils): Get listed buffers and check buffer closability

- For consistency with upstream updates with #8, refactor utils.get_buffer
  -> utils.get_listed_buffers to enable autoclose of listed unnamed buffers
- For consistency in autoclose behavior without repeating logic, extract a
  utils.buf_autoclosable for single source of truth on buffer autoclosability
  with autoclose or close_unpinned.

ditsuke commented on lua/hbac/command/init.lua:
```diff
@@ -0,0 +1,54 @@
+	vim.notify(bufnr .. " " .. pinned_state, "info", notify_opts)
+end
+
+M.subcommands.toggle_all = function()
```
I see a potential problem with `toggle_all` in that it, and correct me if I'm wrong, it doesn't update with the pin status of buffers. Lets say we `toggle_all` to pin all buffers with 1, 2, 3 open and then close a buffer and open some more: 1, 2, 4, 5. What should be the behavior when calling `toggle_all` again? Judging by the current code we're going to unpin all buffers, which might not be ideal.

axkirillov replied:
That's a good point. Maybe the simplest way to deal with this is to have 2 separate commands. Otherwise the logic gets kinda bizzare. With 2 commands the user can expect the same thing every time. Pin all pins all buffers. Unpin all unpins all. Simple to implement and simple for the user to use.

al-ce replied:
Sounds good, I think I had just gotten used to calling the command twice quickly with a convenient sequence but I realize that's kinda hacky. Will split.

ditsuke commented on lua/hbac/setup.lua:
```diff
@@ -2,12 +2,21 @@ local state = require("hbac.state")
 			vim.api.nvim_buf_delete(bufnr, {})
 		end,
+		telescope = {
+			mappings = {
```
This is not clear about normal and insert mappings (I had to see the map function to determine it's `<action> = {<insert_mapping>, <normal_mapping>}`). Can we do it like Telescope's own configuration table?

```lua
mappings = {
	n = {...}
	i = {...}
}
```

al-ce replied:
Sure thing, I agree it's good to be consistent, I was just going for a more compressed opts table for the user, but I realize that's easier for me to visualize as the one writing it.

axkirillov replied:
I agree with ditsuke here. I think the most usual case will be that the user just leaves the table empty and the default mappings will be used. If they want to modify they will be most likely more familiar with the telescope way.

al-ce replied:
So the opts could become:

```lua
  mappings = {
    n = {
      close_unpinned = "<M-c>",
      delete_buffer = "<M-x>",
      toggle_all = "<M-a>",
      toggle_selections = "<M-y>",
    },
    i = {
      close_unpinned = "<M-c>",
      delete_buffer = "<M-x>",
      toggle_all = "<M-a>",
      toggle_selections = "<M-y>",
    },
  },
```

and the mapping loop in the `attach_mappings` module would become (a much nicer looking)

```lua
for mode, hbac_cmds in pairs(hbac_config.telescope.mappings) do
  for hbac_cmd, key in pairs(hbac_cmds) do
    map(mode, key, hbac_telescope_actions[hbac_cmd])
  end
end
```

axkirillov replied:
Fine by me, I only use insert mode anyway 😅

al-ce replied:
Updated in https://github.com/axkirillov/hbac.nvim/pull/5/commits/62ac2facb64f425f383a4a6068e8cededd39fc25

al-ce replied:
Updated in https://github.com/axkirillov/hbac.nvim/pull/5/commits/62ac2facb64f425f383a4a6068e8cededd39fc25

al-ce committed:
feat(commands): Split `toggle_all` into `pin_all` and `unpin_all`

Refactor Telescope mappings to match conventional Telescope action mapping opts

Co-authored-by: Aleksandr Kirillov <32141102+axkirillov@users.noreply.github.com>
Co-authored-by: Tushar <hello@ditsuke.com>

al-ce committed:
feat(telescope): Add filename field to picker entries

This change makes the picker compatible with trouble.nvim actions.
Also changes the default picker icons and refactors the icon config
options (ditsuke).

al-ce commented on lua/hbac/command/subcommands.lua:
```diff
@@ -0,0 +1,48 @@
+	return bufnr, pinned_state
+end
+
+M.toggle_all = function(pin_value)
+	local buflist = utils.get_listed_buffers()
+	for _, bufnr in ipairs(buflist) do
+		state.pinned_buffers[bufnr] = pin_value
+	end
+end
+
+M.pin_all = function()
+	M.toggle_all(true)
+end
+
+M.unpin_all = function()
+	M.toggle_all(false)
+end
```
Is this ok regarding the `toggle_all` issue? The notifications are kept in the `command` module functions that call these functions so that the telescope picker can call these without spamming notifications. I don't know if keeping the name `toggle_all` makes sense, maybe `set_all` is more accurate now, but I didn't want to agonize over it.

axkirillov replied:
Set all is a good choice

axkirillov commented on lua/hbac/command/init.lua:
```diff
@@ -0,0 +1,59 @@
+	subcommands = {},
+}
+
+M.subcommands.close_unpinned = function()
```
If it's possible I would move subcommands to the subcommands.lua file.

al-ce replied:
I wanted to make sure I understood your request correctly before I pushed changes here but I tackled it on another branch

2d1575284956bcc44947410b2ba4a56c39543a5e

The reason for abstracting/encapsulating the subcommands functions in command/init.lua was to print notifications for `:Hbac` commands but not when Telescope actions were executed. This was admittedly messy. In that branch, I changed the following (and we can take these changes à la carte):

- Moved format_filepath and get_devicon to utils (irrelevant to this issue but I was doing some cleanup for some personal features)
- Add a [notify option](https://github.com/axkirillov/hbac.nvim/blob/3e6cde0714996e62dd9f7306033231473ade559c/lua/hbac/setup.lua#L11) to the user opts so that notifications for `:Hbac` commands could be suppressed (defaults to true, i.e. notifications on)
- Add a hbac_notify function to standardize vim.notify across different modules. If opts.notify is set to false, vim.notify won't be called
- During each Telescope action, opts.notify is set to false before each action, then restored to its original value after with a `set_notify` function in utils

The main benefits are that the command module doesn't use a separate abstracted function just to print a notification, subcommands should be easier to maintain, and changing how notifications are handled is done in one place. 

Another approach I tried instead of toggling the notify opt off and on during each telescope action was to pass a parameter to the hbac_notify function to show/suppress notifications, but I ran into some issues with the attach_mappings Telescope module regarding how bufnr is passed to the toggle_pin function. It would have involved some bigger restructuring I was hesitant to make. Also, since adding the notify opt seemed like a nice feature, I thought it would be better to use that check in one place rather than passing an extra parameter in multiple places.

These were all fairly easy quick changes, so if this isn't what you had in mind and you'd rather save these kinds of changes for a different issue/PR, we can come back around to it. Thanks!

(EDIT: pushed a fix for the filepath display)

axkirillov replied:
Nice! How about we merge this and open a pr for that other branch?

al-ce replied:
Sounds good! Thanks again

axkirillov commented:
This looks good to me. I will test it some more today. I am going on vacation this week, so you guys can merge it, or we will merge it next week then.

al-ce commented:
> This looks good to me. I will test it some more today. I am going on vacation this week, so you guys can merge it, or we will merge it next week then.

Thanks for the opportunity to work with you again, and @ditsuke too! I appreciate the learning experience. Enjoy your vacation. I'll squash down to fewer commits at some point.

ditsuke commented:
Thanks for working on this @al-ce. I'm busy because of exams right now so it'll take me around a week to add the picker to my own configuration, lets keep merging on hold for a while but everything LGTM from back when I reviewed and tested it last.

axkirillov commented:
~~Trying to delete a buffer results in a following error~~

Never mind, I was using an old mapping

al-ce commented:
Are either of you having issues with the main hbac branch after upgrading to neovim 0.9.1? Sorry I didn't test this properly, I'm noticing right before I run out for the day. The autoclose thershold isn't being respected and the close_unpinned command gives me this error:

```
Error executing lua callback: Vim:E315: ml_get: invalid lnum: 1
stack traceback:
	[C]: in function 'bufload'
	vim/_editor.lua: in function 'region'
	...ar/neovim/0.9.1/share/nvim/runtime/lua/vim/highlight.lua:35: in function 'range'
	...share/nvim/lazy/nvim-ts-rainbow/lua/rainbow/internal.lua:86: in function 'update_range'
	...share/nvim/lazy/nvim-ts-rainbow/lua/rainbow/internal.lua:194: in function 'cb'
	...1/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:736: in function '_do_callback'
	...1/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:176: in function 'invalidate'
	...1/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:877: in function '_on_detach'
	...r/neovim/0.9.1/share/nvim/runtime/lua/vim/treesitter.lua:73: in function <...r/neovim/0.9.1/share/nvim/runtime/lua/vim/treesitter.lua:69>
	[C]: in function 'nvim_buf_delete'
	...rani/.local/share/nvim/lazy/hbac.nvim/lua/hbac/setup.lua:9: in function 'close_command'
	...ni/.local/share/nvim/lazy/hbac.nvim/lua/hbac/command.lua:13: in function <...ni/.local/share/nvim/lazy/hbac.nvim/lua/hbac/command.lua:8>
	...ni/.local/share/nvim/lazy/hbac.nvim/lua/hbac/command.lua:40: in function 'vim_cmd_func'
	...rani/.local/share/nvim/lazy/hbac.nvim/lua/hbac/setup.lua:41: in function <...rani/.local/share/nvim/lazy/hbac.nvim/lua/hbac/setup.lua:40>
```

I'll check in a minimal Neovim config later and open an issue if it seems like it's not a personal issue. I'm just mentioning this here so we don't move forward with anything Telescope related in case we need to refocus our priorities. Thanks

axkirillov commented:
@al-ce I havent updated yet

axkirillov commented:
~~Hey, looking at the stack traceback I am very confused. There is no command.lua file in your branch. setup.lua:9 is also not a call to close_command. Are you sure you are on correct branch?~~
Sorry, should have noticed earlier that you are talking about main branch. Have you tried this one though?

axkirillov commented:
I think we can still merge the pr. We should add it to readme that it doesnt work on 9.1. My intuition is that most people will not upgrade right away. We can then tackle this issue separately in a separate issue.

ditsuke reviewed (approved):
I finally integrated the picker in my setup and tested the way I use the stock buffer picker. Looks good, although I have a _slight_ preference for this set of icons, not that it needs to be default.
![image](https://github.com/axkirillov/hbac.nvim/assets/72784348/9bf71d11-2a28-4aaf-9d26-5fbe796e2641)

Finally I'd like a `sort_mru` option, will work on a PR once we merge this one

al-ce commented:
> I think we can still merge the pr. We should add it to readme that it doesnt work on 9.1. My intuition is that most people will not upgrade right away. We can then tackle this issue separately in a separate issue.

Sounds good, sorry for the panicky message above, I didn't have time to check if I had everything set up correctly.

al-ce commented:
The issue with 0.9.1 was with conflicting plugins that I didn't disable, working fine with an as-minimal-as-possible config. Sorry for the confusion. (so there's nothing to add to the readme)

@ditsuke thanks for testing. I agree about using sort_mru, that would be a good default option. This was my first time implementing a telescope picker so there's a lot of sensible options I may have missed. I can wait to make the PR for the commands/subcommands modules refactoring until after you make one for that (I'm also not attached to the current default icons).

Thanks everyone!

axkirillov commented:
@ditsuke im also indifferent to which icons, so feel free to make yours the default 😁

ditsuke commented:
Changed the default icons. We're ready for a merge
